### PR TITLE
feature(lsp): TextDocumentEdit support

### PR DESCRIPTION
**Pull request by rgrinberg**
Add a function to apply text document edits and use it to test diff
generation

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

**Comment by rgrinberg** [comment]
Note that this function is not used in ocamllsp. It's only used in the tests but is likely useful to anyone developing an lsp client. `TextEdit.t` lists are interpreted differently than change notifications, so it's useful to have this.

File: lsp/src/text_document.ml [hunk 1]
```diff
@@ -159,3 +164,59 @@ let apply_content_changes ?version t changes =
     | Some version -> { document with version }
   in
   { t with document }
+
+type change =
+  { start : int
+  ; stop : int
```
Comment (ulugbekna) [inline lsp/src/text_document.ml:170]: mention `start` and `stop` are offsets would be useful

File: lsp/src/text_document.ml [hunk 2]
```diff
@@ -159,3 +164,59 @@ let apply_content_changes ?version t changes =
...
+
+module Edit_map = Map.Make (struct
+  type t = int
+
+  let compare = Int.compare
+end)
+
+let add_edit map change =
+  (* TODO check non overlapping property *)
+  Edit_map.update map ~key:change.start ~f:(function
```
Comment (ulugbekna) [inline lsp/src/text_document.ml:182]: how is it that we expect non-overlapping ranges but keep a _multi_-map where the key is the same start offset?

File: lsp/src/text_document.ml [hunk 3]
```diff
@@ -159,3 +164,59 @@ let apply_content_changes ?version t changes =
...
+    simplify_changes acc text find_position changes
+
+let apply_changes text encoding changes =
+  let find_position =
+    match encoding with
+    | `UTF8 -> find_offset_8
+    | `UTF16 -> find_offset_16
+  in
+  let simplified = simplify_changes Edit_map.empty text find_position changes in
+  let pos = ref 0 in
```
Comment (ulugbekna) [inline lsp/src/text_document.ml:202]: I think this keeps an index into `text`? (a comment would be helpful) 

I just don't understand the comment "if this is an insert, it's allowed to increase the position" - if `pos` is an index in `text`, shouldn't it remain unchanged on an insertion?

Also, in `pos := max !pos stop` - can `stop` be ever less than `pos` ?

**Review by ulugbekna — COMMENTED** [review]
I'm a bit lost at the goal of the PR:

Do I understand correctly that the main goal of this PR is to support edits without constructing `TextDocumentContentChangeEvent`s? (modulo it has its own edit application)

Do you think it's a good idea to have two edit application functions - `appy_content_changes` vs `apply_text_document_edits` - and test the latter, while the former is used by ocamllsp? Do we plan to transition to the latter in ocamllsp?

File: lsp/src/text_document.ml (hunk 2 above)
Comment (rgrinberg) [reply lsp/src/text_document.ml:182]: We can change the map to use `(start * stop)` as the key.

**Comment by rgrinberg** [comment]
> Do I understand correctly that the main goal of this PR is to support edits without constructing TextDocumentContentChangeEvents? (modulo it has its own edit application)

The relevant goal for us is to test the `TextEdit.t list` we produce for diffs. But as I said, it's a generally useful function for people writing LSP clients (not us) and our lsp library is supposed to be general purpose.

> Do you think it's a good idea to have two edit application functions - appy_content_changes vs apply_text_document_edits - and test the latter, while the former is used by ocamllsp?

Isn't that what we have? 

> Do we plan to transition to the latter in ocamllsp?

Nope, but we will hopefully find these functions more and more useful as we test workspace edits the server generates more thoroughly.

File: lsp/src/text_document.ml (hunk 1 above)
Comment (rgrinberg) [reply lsp/src/text_document.ml:170]: I'll send another PR to refactor this stuff.

File: lsp/src/text_document.ml (hunk 3 above)
Comment (rgrinberg) [reply lsp/src/text_document.ml:202]: Even though I wrote this code only 2 days ago, I'm already not particularly confident in the implementation.  Feel free to improve it and add tests.

>  if pos is an index in text, shouldn't it remain unchanged on an insertion?

If it's an insert, then we have `start = stop`, which means that `pos := max !pos stop` will not advance past the start position of the edit. This should answer your question.

> Also, in pos := max !pos stop - can stop be ever less than pos ?

Only if we provide an invalid range where `start > end_`

**Pull request merged**